libs/core/langchain_core/runnables/history.py

```diff
@@ -350,6 +343,12 @@ def _exit_history(self, run: Run, config: RunnableConfig) -> None:
         input_val = inputs[self.input_messages_key or "input"]
         input_messages = self._get_input_messages(input_val)
 
+        # If historic messages were prepended to the input messages, remove them to
+        # avoid adding duplicate messages to history.
+        if not self.history_messages_key:
+            historic_messages = config["configurable"]["message_history"].messages
```

Comment:
Is this an extra network request?

Reply:
yea. you're thinking we could store the messages in the config when we initially pull them in _enter_history?

Reply:
If possible or maybe there's some other way to do this?

I think we should merge as is for now for correctness, and proceed with optimization at some point later if possible.

We also have multiple network request with add_message which isn't amazing, but not end of world here since it's low throughput

Reply:
will land bug fix for now and follow up with improvements